### ISSUE TITLE
Camera.project_position() always give wrong results when z_depth is Camera.far

### ISSUE THREAD

**Issue by rmalizia44** [godotengine/godot#46010]
**Godot version:**
3.2.3 stable official

**OS/device including version:**
Windows 10 64 bits

**Issue description:**
When I call Camera.project_position() with Camera.far as z_depth, I always have wrong results, the projected point is always at the center of the screen. To have correct results, I have to call this function with Camera.far-1 or Camera.far+1.

**Steps to reproduce:**
I made a minimal example to reproduce the bug:
```gdscript
extends Spatial

func _input(event: InputEvent) -> void:
	if event is InputEventMouseButton and event.is_pressed():
		calc_far(event.position)

func calc_far(position:Vector2) -> void:
	var camera := get_tree().root.get_camera()
	var far := camera.project_position(position, camera.far)
	$Far.global_transform.origin = far
	print(far)
```

**Minimal reproduction project:**
The project only have a Spatial node with the provided script and 2 children: a camera and a sphere mesh to visualize wrong projected positions.

**Comment by Calinou** [comment]
> The project only have a Spatial node with the provided script and 2 children: a camera and a sphere mesh to visualize wrong projected positions.

Can you upload this minimal reproduction project to make testing quicker? Thanks in advance :slightly_smiling_face:

**Comment by rmalizia44** [comment]
Sure =)
[RayCastBug.zip](https://github.com/godotengine/godot/files/5979359/RayCastBug.zip)

**Comment by dreadpon** [comment]
Can confirm this is still happening in 3.5.
Thank you @rmalizia44, your workaround works perfectly. You just saved me (possibly) hours of debugging...

**Comment by dreadpon** [comment]
This seems to be the culprit. The cause is simply that division by 0 is undefined I suppose?

In 3.5 
https://github.com/godotengine/godot/blob/1ea8b9d2b7da2285e890e087ff2b354e4d458b66/core/math/camera_matrix.cpp#L73-L74

In 4.0
https://github.com/godotengine/godot/blob/18596f6b3dd8a3bda04d0367d37f94c579a8a6e7/core/math/projection.cpp#L263-L264


I don't know if there's an elegant solution, but at the very least it should through an error instead of returning, since projecting onto a far plane seems like a reasonable action on it's own. 
Or perhaps a note in documentation would be enough?

**Comment by dreadpon** [comment]
I don't have much experience with matricies, but if there IS a way to project unto a far plane, it should just jump straight to it bypassing these zero checks. Implicit rounding seems counter intuitive.

Or maybe there's some 'smallest' value that is universally used across the engine. 
And then we can do `z_depth -= smallest_value`

**Comment by rmalizia44** [comment]
It seems to me that if the `Camera.far` value is incremented just for this calculation, when `Camera.far` is equal to `z_depth`, there will be no problems.

**Comment by dreadpon** [comment]
Technically yeah, but what is our basis for choosing the increment value?
You add a whole `1.0`, but for some cases that might be too much and cause undesired behavior.
I suppose something like 0.05 or 0.01 should be safe enough, but still. 

Whatever the solution, I think docs should reference this in some way. 
I, like you, expected a projection straight on the far plane, and would have been surprised if the resulting distance was bigger or smaller than intented.

**Comment by rmalizia44** [comment]
In the tests I made for this method, it doesn't matter the value of `camera.far`, it only can't be the same as the z_depth, so for the purpose of this calculation only, the argument `p_z_far` can always be passed as `p_z_near + 1`, like this:

camera3d.cpp line 373-377 master:
```
if (mode == PROJECTION_ORTHOGONAL) {
  cm.set_orthogonal(size, viewport_size.aspect(), p_z_depth, p_z_depth+1, keep_aspect == KEEP_WIDTH);
} else {
  cm.set_perspective(fov, viewport_size.aspect(), p_z_depth, p_z_depth+1, keep_aspect == KEEP_WIDTH);
}
```

I tested with gdscript and the result is always the same

```
var dist = 100;
camera.far = dist+10;
print(camera.project_position(position, dist))
camera.far = dist-10;
print(camera.project_position(position, dist))
```

**Comment by rmalizia44** [comment]
Internally, in Projection::set_perspective, `deltaZ ` will always be `1`

https://github.com/godotengine/godot/blob/18596f6b3dd8a3bda04d0367d37f94c579a8a6e7/core/math/projection.cpp#L252-L276

**Comment by Flarkk** [comment]
Solved by #98489
